js.go

```diff
@@ -0,0 +1,1175 @@
+	"strings"
+	"time"
+)
+
+// JetStream is the public interface for the JetStream context.
+type JetStream interface {
+	// Publishing messages to JetStream.
+	Publish(subj string, data []byte, opts ...PubOpt) (*PubAck, error)
+	PublishMsg(m *Msg, opts ...PubOpt) (*PubAck, error)
+
+	// Subscribing to messages in JetStream.
+	Subscribe(subj string, cb MsgHandler, opts ...SubOpt) (*Subscription, error)
+	SubscribeSync(subj string, opts ...SubOpt) (*Subscription, error)
+	// Channel versions.
+	ChanSubscribe(subj string, ch chan *Msg, opts ...SubOpt) (*Subscription, error)
+	// QueueSubscribe.
+	QueueSubscribe(subj, queue string, cb MsgHandler, opts ...SubOpt) (*Subscription, error)
+
+	// Management
+	// TODO(dlc) - add more
+	AddStream(cfg *StreamConfig) (*StreamInfo, error)
+	AddConsumer(stream string, cfg *ConsumerConfig) (*ConsumerInfo, error)
+}
+
+// ApiError is included in all API responses if there was an error.
+type ApiError struct {
+	Code        int    `json:"code"`
+	Description string `json:"description,omitempty"`
+}
+
+// ApiResponse is a standard response from the JetStream JSON API
+type ApiResponse struct {
+	Type  string    `json:"type"`
+	Error *ApiError `json:"error,omitempty"`
+}
+
+type AccountInfoResponse struct {
+	ApiResponse
+	*AccountStats
+}
+
+// AccountLimits is for the information about
+type AccountLimits struct {
+	MaxMemory    int64 `json:"max_memory"`
+	MaxStore     int64 `json:"max_storage"`
+	MaxStreams   int   `json:"max_streams"`
+	MaxConsumers int   `json:"max_consumers"`
+}
+
+// AccountStats returns current statistics about the account's JetStream usage.
+type AccountStats struct {
+	Memory  uint64        `json:"memory"`
+	Store   uint64        `json:"storage"`
+	Streams int           `json:"streams"`
+	Limits  AccountLimits `json:"limits"`
+}
+
+const defaultRequestWait = 5 * time.Second
+
+// Internal struct for jetstream
+type js struct {
+	nc *Conn
+	// For import JS from other accounts.
+	pre string
+}
+
+// Request API subjects for JetStream.
+const (
+	// JSApiAccountInfo is for obtaining general information about JetStream.
+	JSApiAccountInfo = "$JS.API.INFO"
+	// JSApiStreams can lookup a stream by subject.
+	JSApiStreams = "$JS.API.STREAM.NAMES"
+	// JSApiConsumerCreateT is used to create consumers.
+	JSApiConsumerCreateT = "$JS.API.CONSUMER.CREATE.%s"
+	// JSApiDurableCreateT is used to create durable consumers.
+	JSApiDurableCreateT = "$JS.API.CONSUMER.DURABLE.CREATE.%s.%s"
+	// JSApiConsumerInfoT is used to create consumers.
+	JSApiConsumerInfoT = "$JS.API.CONSUMER.INFO.%s.%s"
+	// JSApiRequestNextT is the prefix for the request next message(s) for a consumer in worker/pull mode.
+	JSApiRequestNextT = "$JS.API.CONSUMER.MSG.NEXT.%s.%s"
+	// JSApiStreamCreate is the endpoint to create new streams.
+	JSApiStreamCreateT = "$JS.API.STREAM.CREATE.%s"
+)
+
+// JetStream returns a JetStream context for pub/sub interactions.
+func (nc *Conn) JetStream(optApiPre ...string) (JetStream, error) {
```

Comment:
seems we'd want to pass other options in here like default timeout etc, this should probably be a list of functions

Reply:
Yes I agree with you, was trying to set the stage quickly for multiple contexts but server needs a few more minor mods for it to fully work.

Reply:
ok made this change and squashed.